src/Typin/Typin/Experimental/Design.cs

```diff
@@ -0,0 +1,174 @@
+
+        /// <summary>
+        /// A short summary able to be printed in a single console line, when the command is presented 
+        /// in a list of other sibling commands
+        /// </summary>
+        public string Summary { get; init; }
+
+        /// <summary>
+        /// A description, possibly more extense than the <see cref="Summary"/> decribing what the command does
+        /// </summary>
+        public string Description { get; init; }
+
+        /// <summary>
+        /// List of argument descriptors specifying the arguments of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandArgumentDescriptor> Arguments { get; init; }
+
+        /// <summary>
+        /// List of option descriptors specifying the options of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandOptionDescriptor> Options { get; init; }
+
+        /// <summary>
+        /// List of examples to show to the user for current command
+        /// </summary>
+        public IReadOnlyCollection<CommandExampleDescriptor> Examples { get; init; }
+    }
+
+    public class CommandArgumentDescriptor
+    {
+        public string Name { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public int Position { get; init; }
+        public bool Multiple { get; init; }
+    }
+
+    public enum OptionKind
+    {
+        Flag,
+        Value,
+    }
+
+    public class CommandOptionDescriptor
+    {
+        public string Name { get; init; }
+        public IReadOnlyCollection<string> ShortAliases { get; init; }
+
+        /// <summary>
+        /// For flags, you can specify a negative name, like no-verbose, no-build, etc., so the user can
+        /// explicitly indicate that he wants to give a false value for the flag, disregarding it's default value
+        /// </summary>
+        public string NegativeName { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public bool Multiple { get; init; }
+        public OptionKind OptionKind { get; init; }
+        public string DefaultValue { get; init; }
+    }
+
+    public class CommandExampleDescriptor
+    {
+        public string Description { get; init; }
+    }
+
+    // Pure Base Runtime
+
+    public interface ICommandRunner
+    {
+        ValueTask ExecuteAsync(CommandRunnerContext context, CancellationToken cancellationToken);
+    }
+
+    public class CommandRunnerContext
+    {
+        public ICliContext CliContext { get; init; }
+        public CommandDescriptor CommandDescriptor { get; init; }
+        public string CommandLine { get; init; }
+        public IReadOnlyCollection<ParsedArgument> ParsedArguments { get; init; }
+        public IReadOnlyCollection<ParsedOption> ParsedOptions { get; init; }
+    }
+
+    public class ParsedArgument
+    {
+        public CommandArgumentDescriptor Descriptor { get; init; }
```

Comment:
That's why I created the class `CommandRunnerContext` which is the context in which current command is running, and that context is prepared after converting the `IEnumerable<string>`, using the CommandDescriptor that corresponds to the typed command, into a parsed command, which know the `CommandDescriptor` and all the `CommandParameterDescriptor` and `CommandOptionDescriptor` used to parse the command line.

So the `ICommandRunner`, instead of receiving a `ICliContext`, would receive a more specialized context with more information after the parsing have happened.

Reply:
`That's why I created the class `CommandRunnerContext` (...)` - `ICliContext`/`CliContext` does exactly the same :)

Reply:
Well, maybe not exactly. We have just a raw or partially parsed input - CliContext lack a relation between input and ICommand arguments (some intermediate output from binder).

Reply:
I tried to follow the basics of ASP.NET filters, where you have access to `HttpContext`, but also to a specific context to the filter at hand, where you have other processed data. like thrown exception, `ActionDescriptor`, etc.

So you have `HttpContext` (like ICliContext) and `ExceptionContext`, `ActionExecutingContext` and others, representing current command execution (with parsed arguments, and command metadata).

Maybe we can design it differently, but I just laid out a first approach.